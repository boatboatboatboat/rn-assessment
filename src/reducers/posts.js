import {createAsyncThunk, createSlice} from "@reduxjs/toolkit";
import {api} from "../api";

export const PostStatus = {
    Loading: 0,
    Ready: 1,
    Failed: 2,
};

export const fetchPosts = createAsyncThunk("posts/fetchPosts", async () => {
    return await api.getPosts();
});

export const fetchPost = createAsyncThunk("posts/fetchPost", async postId => {
    return await api.getPostById(postId);
});

export const slice = createSlice({
    name: "posts",
    initialState: {
        posts: [],
        status: PostStatus.Loading,
    },
    reducers: {},
    extraReducers: {
        [fetchPost.fulfilled]: (state, action) => {
            const newPost = action.payload;
            const postId = action.meta.arg;
            const postIndex = state.posts.findIndex(x => x.id === postId);
            if (postIndex === -1) {
                // The post is not in our store, so add it.
                state.posts.push(newPost);
            } else {
                // We already have the post, so update it instead.
                state.posts[postIndex] = newPost;
            }
        },
        [fetchPosts.pending]: state => {
            state.status = PostStatus.Loading;
        },
        [fetchPosts.fulfilled]: (state, action) => {
            state.status = PostStatus.Ready;
            state.posts = action.payload;
        },
        [fetchPosts.rejected]: state => {
            state.status = PostStatus.Failed;
            state.posts = [];
        },
    },
});
export default slice.reducer;
