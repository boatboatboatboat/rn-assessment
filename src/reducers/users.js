import {createAsyncThunk, createSlice} from "@reduxjs/toolkit";
import api from "../api";

export const UserLoadStatus = {
    Loading: 0,
    Ready: 1,
    Failed: 2,
};

export const fetchUser = createAsyncThunk(
    "users/fetchUser",
    async (userId, thunkAPI) => {
        // Check whether the user is being fetched already,
        // if they are, do not fetch again.
        // If the user does already exist, do fetch again.
        const users = thunkAPI.getState().users.users;
        const user = users.find(x => x.id === userId);
        if (user != null) {
            const isCurrentRequester =
                thunkAPI.requestId === user.loadStatus.requestId;

            if (
                user.loadStatus.status !== UserLoadStatus.Ready &&
                !isCurrentRequester
            ) {
                // It's already being fetched, so don't fetch.
                return;
            }
        }
        return await api.getUserById(userId);
    },
);

export const slice = createSlice({
    name: "users",
    initialState: {
        users: [],
    },
    extraReducers: {
        [fetchUser.pending]: (state, action) => {
            const userId = action.meta.arg;
            const user = state.users.find(x => x.id === userId);

            if (user == null) {
                // If the user hasn't been added to the store yet,
                // create the user in the store and set the loading state.
                state.users.push({
                    id: userId,
                    loadStatus: {
                        status: UserLoadStatus.Loading,
                        requestId: action.meta.requestId,
                    },
                });
            } else if (user.loadStatus.status !== UserLoadStatus.Loading) {
                // If the user has been added to the store before,
                // change the existing user's state into loading.
                user.loadStatus = {
                    status: UserLoadStatus.Loading,
                    requestId: action.meta.requestId,
                };
            }
        },
        [fetchUser.rejected]: (state, action) => {
            const {requestId, arg: userId} = action.meta;
            const user = state.users.find(x => x.id === userId);

            // Ignore the errors from inactive requests
            if (user != null && user.loadStatus.requestId === requestId) {
                user.loadStatus.status = UserLoadStatus.Failed;
            }
        },
        [fetchUser.fulfilled]: (state, action) => {
            const {requestId, arg: userId} = action.meta;
            const userIndex = state.users.findIndex(x => x.id === userId);
            const user = state.users[userIndex];

            // The payload of an inactive request is null,
            // so we should ignore them.
            if (user != null && user.loadStatus.requestId === requestId) {
                state.users[userIndex] = action.payload;
                state.users[userIndex].loadStatus = {
                    status: UserLoadStatus.Ready,
                };
            }
        },
    },
});

export default slice.reducer;
