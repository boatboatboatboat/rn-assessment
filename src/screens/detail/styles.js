import {StyleSheet} from "react-native";
import {DarkColors, DefaultColors} from "../../styles/colors";

export function getStyles(theme) {
    const Colors = theme === "dark" ? DarkColors : DefaultColors;
    return StyleSheet.create({
        header: {
            paddingHorizontal: 16,
            paddingTop: 16,
        },
        title: {
            color: Colors.foreground,
            fontSize: 20,
            fontWeight: "bold",
        },
        author: {
            color: Colors.foreground,
        },
        body: {
            color: Colors.foreground,
            paddingHorizontal: 16,
        },
        separator: {
            backgroundColor: Colors.border,
            height: 1,
            width: "95%",
            alignSelf: "center",
            marginVertical: 8,
        },
        errorMessage: {
            color: Colors.foreground,
            textAlign: "center",
        },
        errorBackground: {
            backgroundColor: Colors.background,
            borderTopWidth: 0.5,
            borderColor: Colors.border,
            flexGrow: 1,
        },
        background: {
            backgroundColor: Colors.background,
            borderTopWidth: 0.5,
            borderColor: Colors.border,
        },
    });
}

export default getStyles;
