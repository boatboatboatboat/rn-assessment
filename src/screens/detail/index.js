import React, {useEffect} from "react";
import {ScrollView, Text, View, useColorScheme} from "react-native";
import {useDispatch, useSelector} from "react-redux";
import {fetchUser} from "../../reducers/users";
import {fetchPost, PostStatus} from "../../reducers/posts";
import getStyles from "./styles";

export function DetailedPost({route}) {
    const styles = getStyles(useColorScheme());
    const dispatch = useDispatch();
    const {postId} = route.params;
    const post = useSelector(state =>
        state.posts.posts.find(x => x.id === postId),
    );
    const user = useSelector(state =>
        state.users.users.find(x => post != null && x.id === post.userId),
    );

    useEffect(() => {
        // Fetch the user if they aren't in our store yet.
        if (post == null) {
            dispatch(fetchPost(postId));
        } else if (user == null) {
            dispatch(fetchUser(post.userId));
        }
    });

    if (post == null) {
        // If for whatever reason we can't load the post,
        // show an error message.
        // This should never happen though.
        return (
            <View style={styles.errorBackground}>
                <Text style={styles.errorMessage}>
                    Failed to load the post! (id: {postId})
                </Text>
            </View>
        );
    }

    const authorInfo = () => {
        // Only render the author if the user has been fetched.
        if (user != null && user.loadStatus.status === PostStatus.Ready) {
            return (
                <Text style={styles.author}>
                    by {user.name} ({user.username})
                </Text>
            );
        }
    };

    return (
        <ScrollView style={styles.background}>
            <View style={styles.header}>
                <Text style={styles.title}>{post.title}</Text>
                {authorInfo()}
            </View>
            <View style={styles.separator} />
            <Text style={styles.body}>{post.body}</Text>
        </ScrollView>
    );
}
