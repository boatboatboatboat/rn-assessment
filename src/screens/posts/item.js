import React, {useEffect} from "react";
import {Text, View, useColorScheme} from "react-native";
import {useDispatch, useSelector} from "react-redux";
import {fetchUser, UserLoadStatus} from "../../reducers/users";
import getStyles from "./styles";

export function PostItem(props) {
    const styles = getStyles(useColorScheme());
    const dispatch = useDispatch();
    const post = useSelector(state =>
        state.posts.posts.find(x => x.id === props.postId),
    );
    const user = useSelector(state =>
        state.users.users.find(x => post != null && x.id === post.userId),
    );

    // If we can't find the user in our store,
    // try and fetch the user.
    useEffect(() => {
        if (post != null && user == null) {
            dispatch(fetchUser(post.userId));
        }
    });

    if (post == null) {
        // Don't render anything if the postId is invalid
        return <View />;
    }

    const authorText = () => {
        // Only show the author when we've (successfully) fetched the user.
        if (user != null && user.loadStatus.status === UserLoadStatus.Ready) {
            return <Text style={styles.itemAuthor}>by {user.name}</Text>;
        }
    };

    return (
        <View style={styles.itemBackground}>
            <Text style={styles.itemTitle}>{post.title}</Text>
            {authorText()}
        </View>
    );
}
