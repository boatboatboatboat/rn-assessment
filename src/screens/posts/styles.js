import {StyleSheet} from "react-native";
import {DarkColors, DefaultColors} from "../../styles/colors";

export function getStyles(theme) {
    const Colors = theme === "dark" ? DarkColors : DefaultColors;

    return StyleSheet.create({
        listBackground: {
            flexGrow: 1,
            backgroundColor: Colors.listBackground,
            borderTopWidth: 0.5,
            borderColor: Colors.border,
        },
        listStatusMessage: {
            color: Colors.foreground,
            marginVertical: 1,
            padding: 16,
            textAlign: "center",
        },
        listItemSeparator: {
            backgroundColor: Colors.border,
            height: 1,
            width: "100%",
        },
        itemBackground: {
            backgroundColor: Colors.background,
            padding: 16,
        },
        itemPressedOverlay: {
            ...StyleSheet.absoluteFillObject,
            backgroundColor: "rgba(255,255,255,0.2)",
        },
        itemTitle: {
            color: Colors.foreground,
            fontSize: 20,
        },
        itemAuthor: {
            color: Colors.foreground,
        },
    });
}

export default getStyles;
