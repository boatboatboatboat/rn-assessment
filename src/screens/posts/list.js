import React from "react";
import {FlatList, Pressable, Text, View, useColorScheme} from "react-native";
import {useDispatch, useSelector} from "react-redux";
import {fetchPosts, PostStatus} from "../../reducers/posts";
import {PostItem} from "./item";
import getStyles from "./styles";

export function PostList({navigation}) {
    const styles = getStyles(useColorScheme());
    const dispatch = useDispatch();
    const posts = useSelector(state => state.posts.posts);
    const status = useSelector(state => state.posts.status);

    const getStatusMessage = () => {
        if (status === PostStatus.Loading) {
            return "Loading...";
        } else if (status === PostStatus.Failed) {
            return "Failed to retrieve posts - pull down to retry";
        } else {
            return "There are no posts!";
        }
    };

    const renderPressablePostItem = ({item}) => {
        return (
            <Pressable
                onPress={() =>
                    navigation.navigate("Detailed", {postId: item.id})
                }>
                {({pressed}) => {
                    const postItem = <PostItem postId={item.id} />;
                    // If pressed, highlight the button
                    // by using a bright semitransparent overlay
                    if (pressed) {
                        return (
                            <>
                                {postItem}
                                <View style={styles.itemPressedOverlay} />
                            </>
                        );
                    } else {
                        return postItem;
                    }
                }}
            </Pressable>
        );
    };

    const postsInReverseIdOrder = [...posts].sort((x, y) => y.id > x.id);

    const listItemSeparator = () => {
        return <View style={styles.listItemSeparator} />;
    };

    return (
        <View style={styles.listBackground}>
            <FlatList
                data={postsInReverseIdOrder}
                renderItem={renderPressablePostItem}
                refreshing={status === PostStatus.Loading}
                onRefresh={() => dispatch(fetchPosts())}
                ListEmptyComponent={
                    <Text style={styles.listStatusMessage}>
                        {getStatusMessage()}
                    </Text>
                }
                ItemSeparatorComponent={listItemSeparator}
            />
        </View>
    );
}
