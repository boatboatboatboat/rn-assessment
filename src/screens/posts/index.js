import {PostItem} from "./item";
import {PostList} from "./list";

export {PostList, PostItem};
