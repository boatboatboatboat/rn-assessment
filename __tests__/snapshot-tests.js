import fetchMock from "fetch-mock";
import React from "react";
import {Provider} from "react-redux";
import renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import {PostItem, PostList} from "../src/screens/posts";
import {PostStatus} from "../src/reducers/posts";
import {UserLoadStatus} from "../src/reducers/users";
import {DetailedPost} from "../src/screens/detail";

const mockStore = configureStore();
const store = mockStore({
    posts: {
        posts: [
            {
                id: 42,
                userId: 3,
                title: "THIS IS a TITLE",
                body: "check getPosts url",
            },
            {
                id: 4,
                userId: 5,
                title: "hello!",
                body: "this is a body",
            },
        ],
        status: PostStatus.Ready,
    },
    users: {
        users: [
            {
                id: 3,
                name: "Test name",
                username: "Testusername",
                loadStatus: {status: UserLoadStatus.Ready},
            },
            {
                id: 5,
                name: "Other tester",
                username: "Usertest",
                loadStatus: {status: UserLoadStatus.Ready},
            },
        ],
    },
});

describe("Snapshot tests", () => {
    afterEach(() => {
        fetchMock.restore();
    });

    it("Simple item snapshot with valid id", () => {
        const tree = renderer.create(
            <Provider store={store}>
                <PostItem postId={4} />
            </Provider>,
        );
        expect(tree.toJSON()).toMatchSnapshot();
    });

    it("Simple item snapshot with invalid id", () => {
        const tree = renderer.create(
            <Provider store={store}>
                <PostItem postId={0} />
            </Provider>,
        );
        expect(tree.toJSON()).toMatchSnapshot();
    });

    it("Posts list snapshot", () => {
        const tree = renderer.create(
            <Provider store={store}>
                <PostList />
            </Provider>,
        );
        expect(tree.toJSON()).toMatchSnapshot();
    });

    it("Detailed post snapshot", () => {
        const tree = renderer.create(
            <Provider store={store}>
                <DetailedPost route={{params: {postId: 4}}} />
            </Provider>,
        );
        expect(tree.toJSON()).toMatchSnapshot();
    });
});
