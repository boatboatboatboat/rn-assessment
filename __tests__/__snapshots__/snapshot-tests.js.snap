// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot tests Detailed post snapshot 1`] = `
<RCTScrollView
  style={
    Object {
      "backgroundColor": "rgb(255, 255, 255)",
      "borderColor": "rgb(216, 216, 216)",
      "borderTopWidth": 0.5,
    }
  }
>
  <View>
    <View
      style={
        Object {
          "paddingHorizontal": 16,
          "paddingTop": 16,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "rgb(28, 28, 30)",
            "fontSize": 20,
            "fontWeight": "bold",
          }
        }
      >
        hello!
      </Text>
      <Text
        style={
          Object {
            "color": "rgb(28, 28, 30)",
          }
        }
      >
        by 
        Other tester
         (
        Usertest
        )
      </Text>
    </View>
    <View
      style={
        Object {
          "alignSelf": "center",
          "backgroundColor": "rgb(216, 216, 216)",
          "height": 1,
          "marginVertical": 8,
          "width": "95%",
        }
      }
    />
    <Text
      style={
        Object {
          "color": "rgb(28, 28, 30)",
          "paddingHorizontal": 16,
        }
      }
    >
      this is a body
    </Text>
  </View>
</RCTScrollView>
`;

exports[`Snapshot tests Posts list snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "rgb(242, 242, 242)",
      "borderColor": "rgb(216, 216, 216)",
      "borderTopWidth": 0.5,
      "flexGrow": 1,
    }
  }
>
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    ListEmptyComponent={
      <Text
        style={
          Object {
            "color": "rgb(28, 28, 30)",
            "marginVertical": 1,
            "padding": 16,
            "textAlign": "center",
          }
        }
      >
        There are no posts!
      </Text>
    }
    data={
      Array [
        Object {
          "body": "check getPosts url",
          "id": 42,
          "title": "THIS IS a TITLE",
          "userId": 3,
        },
        Object {
          "body": "this is a body",
          "id": 4,
          "title": "hello!",
          "userId": 5,
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onRefresh={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        refreshing={false}
      />
    }
    refreshing={false}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <RCTRefreshControl />
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
        >
          <View
            style={
              Object {
                "backgroundColor": "rgb(255, 255, 255)",
                "padding": 16,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "rgb(28, 28, 30)",
                  "fontSize": 20,
                }
              }
            >
              THIS IS a TITLE
            </Text>
            <Text
              style={
                Object {
                  "color": "rgb(28, 28, 30)",
                }
              }
            >
              by 
              Test name
            </Text>
          </View>
        </View>
        <View
          style={
            Object {
              "backgroundColor": "rgb(216, 216, 216)",
              "height": 1,
              "width": "100%",
            }
          }
        />
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
        >
          <View
            style={
              Object {
                "backgroundColor": "rgb(255, 255, 255)",
                "padding": 16,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "rgb(28, 28, 30)",
                  "fontSize": 20,
                }
              }
            >
              hello!
            </Text>
            <Text
              style={
                Object {
                  "color": "rgb(28, 28, 30)",
                }
              }
            >
              by 
              Other tester
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Snapshot tests Simple item snapshot with invalid id 1`] = `<View />`;

exports[`Snapshot tests Simple item snapshot with valid id 1`] = `
<View
  style={
    Object {
      "backgroundColor": "rgb(255, 255, 255)",
      "padding": 16,
    }
  }
>
  <Text
    style={
      Object {
        "color": "rgb(28, 28, 30)",
        "fontSize": 20,
      }
    }
  >
    hello!
  </Text>
  <Text
    style={
      Object {
        "color": "rgb(28, 28, 30)",
      }
    }
  >
    by 
    Other tester
  </Text>
</View>
`;
